README.md
# Caylir

Cayley driver for Elixir

## Cayley Support

Tested Cayley versions:

- `0.7.3`
- `0.7.4`
- `0.7.5`
- `0.7.6`

## Package Setup

To use Caylir with your projects, edit your `mix.exs` file and add the required dependencies:

```elixir
defp deps do
  [
    # ...
    {:caylir, "~> 1.0"},
    # ...
  ]
end
```

## Application Setup

### Graph Definition

Defining a graph connection requires defining a module:

```elixir
defmodule MyGraph do
  use Caylir.Graph, otp_app: :my_app
end
```

This defined graph module needs to be hooked up into your supervision tree:

```elixir
children = [
  # ...
  MyGraph,
  # ...
]
```

For a more detailed explanation of how to get started with a graph please consult the inline documentation of the `Caylir.Graph` module.

To configure your connection you can use the application environment:

```elixir
config :my_app, MyGraph,
  host: "cayley.host",
  port: 42160,
  scheme: "https"
```

The entry should match the chosen `:otp_app` and module name defined earlier.

For more information on how (and what) to configure please refer to the inline documentation of the `Caylir.Graph.Config` module.

## Usage

Querying data:

```elixir
MyGraph.query("graph.Vertex('graph').Out('connection').All()")
```

Writing Data:

```elixir
MyGraph.write(%{
  subject: "graph",
  predicate: "connection",
  object: "target"
})
```

Deleting Data:

```elixir
MyGraph.delete(%{
  subject: "graph",
  predicate: "connection",
  object: "target"
})
```

A more detailed usage documentation can be found inline at the `Caylir.Graph` module.

## License

[Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0)


CHANGELOG.md
# Changelog

## v2.0.0-dev

- Backwards incompatible changes
    - Minimum required Elixir version is now `~> 1.11`

## v1.0.0 (2020-03-25)

- Enhancements
    - A missing `host` configuration will now default to `"localhost"`
    - A missing `port` configuration will now default to `64210`
    - A missing `otp_app` configuration will no longer raise and only use the (inline) defaults
    - Documentation is now available inline (`@moduledoc`, ...) with the `README.md` file targeting the repository (development) instead of releases
    - Errors from `:hackney.body/1` are now passed to the caller instead of raising a `MatchError`
    - Queries are executed directly in the calling process without passing through a process pool
    - The scheme used for the requests can now be configured

- Bug fixes
    - Shape queries now accept a `:language` option as expected

- Backwards incompatible changes
    - Minimum required Elixir version is now `~> 1.7`
    - Support for Cayley versions earlier than `0.7.3` is no longer guaranteed

## v0.11.1 (2019-11-20)

- Bug fixes
    - Decoding json with string keys now works as intended ([#2](https://github.com/mneudert/caylir/pull/2))

## v0.11.0 (2019-04-28)

- Enhancements
    - Initializer modules can be defined with additional arguments by using `{mod, fun, args}`
    - The JSON encoder/decoder can now be defined using a `{mod, fun}` tuple

- Backwards incompatible changes
    - The default JSON encoder/decoder has been switched from `:poison` to `:jason`

## v0.10.0 (2018-10-28)

- Enhancements
    - The library used for JSON encoding/decoding can be changed by using the `:json_encoder` and `:json_decoder` configuration

- Backwards incompatible changes
    - Support for `{:system, "ENV_VARIABLE"}` configuration has been removed

## v0.9.0 (2018-09-30)

- Enhancements
    - Compiling a graph module without passing an `:otp_app` will now raise
    - Graphs can be declared with compile time configuration defaults that are later overwritten by the application environment

- Backwards incompatible changes
    - Minimum required Elixir version is now `~> 1.5`
    - Public access to the internal pool child spec has been removed
    - Public access to the internal pool module name has been removed
    - Support for Cayley version `0.6.1` has been dropped

- Deprecations
    - Accessing the system environment by configuring `{:system, var}` or `{:system, var, default}` will now result in a `Logger.info/1` message and will stop working in a future release

## v0.8.0 (2018-09-09)

- Enhancements
    - Configuration can be done on graph (re-) start by setting a `{mod, fun}` tuple for the config key `:init`. This method will be called with the graph module name as the first (and only) parameter and is expected to return `:ok`
    - Queries now support the `limit` parameter
    - Support for Elixir 1.5 style `child_spec` has been added

- Soft deprecations (no warnings)
    - Support for `{:system, "ENV_VARIABLE"}` configuration has been removed from the documentation. It will eventually be removed completely after a proper deprecation phase

## v0.7.0 (2017-12-28)

- Enhancements
    - Basic support for defining a default graph query language

- Backwards incompatible changes
    - Default query endpoint has been updated to support Cayley version `0.7.0`. To use an older version you need to add `language: :gremlin` to your graph configuration
    - Support for Cayley version `0.6.0` has been dropped

## v0.6.0 (2017-09-17)

- Backwards incompatible changes
    - Minimum required Elixir version is now `~> 1.3`

## v0.5.0 (2017-09-09)

- Enhancements
    - Configuration values can be fetched from the system environment using `{:system, ENV_VAR}` or `{:system, ENV_VAR, default}`

## v0.4.0 (2017-05-18)

- Enhancements
    - Bulk operations for write/delete are now supported

## v0.3.0 (2016-12-11)

- Backwards incompatible changes
    - Minimum required Elixir version is now `~> 1.2`
    - Minimum required Erlang version is now `~> 18.0`
    - Support for `:poison < 2.0` has been removed

## v0.2.0 (2015-06-13)

- Enhancements
    - Allows retrieving the shape of a query
    - Pools connections to the graph
    - Uses `:hackney` instead of `:inets`

- Backwards incompatible changes
    - Configuration is done using config files

## v0.1.0 (2015-06-04)

- Initial Release


mix.exs
defmodule Caylir.MixProject do
  use Mix.Project

  @url_changelog "https://hexdocs.pm/caylir/changelog.html"
  @url_github "https://github.com/mneudert/caylir"
  @version "2.0.0-dev"

  def project do
    [
      app: :caylir,
      name: "Caylir",
      version: @version,
      elixir: "~> 1.11",
      deps: deps(),
      description: "Cayley driver for Elixir",
      dialyzer: dialyzer(),
      docs: docs(),
      package: package(),
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test
      ],
      test_coverage: [tool: ExCoveralls]
    ]
  end

  def application, do: []

  defp deps do
    [
      {:credo, "~> 1.6", only: :dev, runtime: false},
      {:dialyxir, "~> 1.2", only: :dev, runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:excoveralls, "~> 0.15.0", only: :test, runtime: false},
      {:hackney, "~> 1.6"},
      {:jason, "~> 1.0"},
      {:ranch, "~> 1.7.0", only: :test}
    ]
  end

  defp dialyzer do
    [
      flags: [
        :error_handling,
        :underspecs,
        :unmatched_returns
      ],
      plt_core_path: "plts",
      plt_local_path: "plts"
    ]
  end

  defp docs do
    [
      main: "Caylir.Graph",
      extras: [
        "CHANGELOG.md",
        LICENSE: [title: "License"],
        "README.md": [title: "Overview"]
      ],
      formatters: ["html"],
      source_ref: "v#{@version}",
      source_url: @url_github
    ]
  end

  defp package do
    %{
      files: ["CHANGELOG.md", "LICENSE", "mix.exs", "README.md", "lib"],
      licenses: ["Apache-2.0"],
      links: %{
        "Changelog" => @url_changelog,
        "GitHub" => @url_github
      }
    }
  end
end
